File: ROBOTcollides.cpp
class Solution {
public:
    vector<int> survivedRobotsHealths(vector<int>& positions, vector<int>& healths, string directions) {
        int n = positions.size();
        vector<int> indices(n);
        for (int i = 0; i < n; i++) {
            indices[i] = i;
        }
        sort(indices.begin(), indices.end(), [&positions](int a, int b) {
            return positions[a] < positions[b];
        });
        
        vector<int> stack;
        for (int i : indices) {
            if (directions[i] == 'R') {
                stack.push_back(i);
            } else {
                while (!stack.empty() && healths[stack.back()] < healths[i]) {
                    healths[i] -= 1;
                    healths[stack.back()] = 0;
                    stack.pop_back();
                }
                if (!stack.empty()) {
                    if (healths[stack.back()] == healths[i]) {
                        healths[stack.back()] = 0;
                        stack.pop_back();
                    } else {
                        healths[stack.back()] -= 1;
                    }
                    healths[i] = 0;
                }
            }
        }
        
        vector<int> result;
        for (int health : healths) {
            if (health > 0) {
                result.push_back(health);
            }
        }
        
        return result;
    }
};

File: ROBOTcollides.java
class Solution {
    public List<Integer> survivedRobotsHealths(int[] positions, int[] healths, String directions) {
        int n = positions.length;
        Integer[] indexes = new Integer[n];

        for (int i = 0; i < n; i++) {
            indexes[i] = i;
        }

        Arrays.sort(indexes, (a, b) -> positions[a] - positions[b]);

        List<Integer> res = new ArrayList<>();
        Stack<Integer> stack = new Stack<>();

        for (int index : indexes) {
            if (directions.charAt(index) == 'R') {
                stack.push(index);
            } else {
                while (!stack.isEmpty() && healths[index] > 0) {
                    int topIndex = stack.peek();
                    if (directions.charAt(topIndex) == 'R') {
                        if (healths[topIndex] == healths[index]) {
                            healths[topIndex] = 0;
                            healths[index] = 0;
                            stack.pop();
                        } else if (healths[topIndex] < healths[index]) {
                            healths[topIndex] = 0;
                            healths[index]--;
                            stack.pop();
                        } else {
                            healths[topIndex]--;
                            healths[index] = 0;
                        }
                    } else {
                        break;
                    }
                }
                if (healths[index] > 0) {
                    stack.push(index);
                }
            }
        }

        for (int health : healths) {
            if (health > 0) {
                res.add(health);
            }
        }

        return res;
    }
}
